symbolic-base/src/ZkFold/Base/Data/Logical.hs

```diff
@@ -0,0 +1,220 @@
+{-# LANGUAGE DerivingVia          #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module ZkFold.Base.Data.Logical
+  ( Boolean (..)
+  , all, any, and, or
+  , Conditional (..)
+  , ifThenElse
+  , Eq (..)
+  , elem
+  , GConditional (..)
+  , GEq (..)
+  ) where
+
+import           Control.Category
+import qualified Data.Bool                        as H
+import           Data.Foldable                    hiding (all, and, any, elem, or)
+import           Data.Functor.Identity
+import qualified Data.Functor.Rep                 as R
+import           Data.Kind
+import qualified GHC.Generics                     as G
+import           Prelude                          (type (~))
+import qualified Prelude                          as H
+
+import           ZkFold.Base.Algebra.Basic.Number
+import           ZkFold.Base.Data.Vector
+
+class Boolean b where
+
+  true, false :: b
+
+  not :: b -> b
+
+  (&&), (||), xor :: b -> b -> b
+
+instance Boolean H.Bool where
+  true = H.True
+  false = H.False
+  not = H.not
+  (&&) = (H.&&)
+  (||) = (H.||)
+  xor = (H./=)
+
+all :: (Boolean b, Foldable t) => (x -> b) -> t x -> b
```

Comment:
I think it would be better to remove `BoolType` and things which depend on it in this same PR to avoid dead code and code duplication in the repo

Reply:
This is a draft PR while I’m working on it. Don’t worry I will remove duplicate code.